dallinger/experiment.py

```diff
@@ -977,117 +984,127 @@ def monitoring_statistics(self, **kw):
-                .all()
-            ]
-            # We don't filter participants because they aren't directly connected to specific networks
-            jparticipants = [
-                n.__json__() for n in Participant.query.order_by(Participant.id).all()
+        if polymorphic_identity is None and "type" in table.columns:
+            observed_types = [
+                r.type for r in db.session.query(table.columns.type).distinct().all()
             ]
-            jtransformations = []
-            if kw.get("transformations"):
-                jtransformations = [
-                    n.__json__()
-                    for n in Transformation.query.filter(
-                        Transformation.network_id.in_(network_ids)
-                    )
-                    .order_by(Transformation.id)
-                    .all()
-                ]
-
-            jvectors = [
-                {
-                    "origin_id": v.origin_id,
-                    "destination_id": v.destination_id,
-                    "id": v.id,
-                    "failed": v.failed,
-                }
-                for v in Vector.query.filter(Vector.network_id.in_(network_ids))
-                .order_by(Vector.id)
-                .all()
+            obj_by_type = [
+                self.pull_table(
+                    table,
+                    polymorphic_identity=_type,
+                    network_roles=network_roles,
+                    network_ids=network_ids,
+                    cls_filter=cls_filter,
+                )
+                for _type in observed_types
             ]
+            return [obj for sublist in obj_by_type for obj in sublist]
 
-        return {
-            "networks": jnetworks,
-            "nodes": jnodes,
-            "vectors": jvectors,
-            "infos": jinfos,
-            "participants": jparticipants,
-            "trans": jtransformations,
-        }
+        if polymorphic_identity is None:
+            cls = get_mapped_class(table)
+        else:
+            assert "type" in table.columns
+            cls = get_polymorphic_mapping(table)[polymorphic_identity]
+
+        if cls_filter is not None and not cls_filter(cls):
+            return
+
+        query = cls.query
+
+        if polymorphic_identity is not None:
+            query = query.filter(cls.type == polymorphic_identity)
+
+        if network_roles is not None:
+            query = query.filter(Network.role.in_(network_roles))
+
+        if network_ids is not None:
+            try:
+                query = query.filter(Network.id.in_(network_ids))
+            except Exception:
+                raise
+
+        if network_roles is not None or network_ids is not None:
+            if "network_id" in table.columns:
+                query = query.join(Network, cls.network_id == Network.id)
+
+        primary_keys = [c.name for c in table.primary_key.columns]
+
+        return query.order_by(*primary_keys).options(undefer("*")).all()
```

Comment:
What happens if there are multiple `primary_keys` can that be the case?

Reply:
Yep, this code will order by all possible primary keys, if they exist. I don't think there are any examples in Dallinger/PsyNet yet though.

Reply:
Ok, in this case, the order will matter. But since this doesn't exist in PsyNet and Dallinger I wouldn't worry about it.